compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt

```diff
@@ -0,0 +1,181 @@
+    private val endLabel = Label()
+    private var defaultLabel = endLabel
+    private lateinit var elseExpression: IrExpression
+    private val cases = ArrayList<Pair<Int, Label>>()
+    private val thenExpressions = ArrayList<Pair<IrExpression, Label>>()
+    private var subject: IrGetValue? = null
+
+    private val mv = codegen.mv
+
+    // @return true if the IrWhen can be emitted as lookupswitch or tableswitch.
+    fun match(): Boolean {
+        for (branch in expression.branches) {
+            if (branch is IrElseBranch) {
+                elseExpression = branch.result
+                defaultLabel = Label()
+            } else {
+                val thenLabel = Label()
+                thenExpressions.add(Pair(branch.result, thenLabel))
+                if (!matchConditions(branch.condition, thenLabel))
+                    return false
+            }
+        }
+
+        // IF is more compact when there are only 1 or fewer branches, in addition to else.
+        if (cases.size <= 1)
+            return false
+
+        return true
+    }
+
+    // psi2ir lowers multiple cases to nested conditions. For example,
+    //
+    // when (subject) {
+    //   a, b, c -> action
+    // }
+    //
+    // is lowered to
+    //
+    // if (if (subject == a)
+    //       true
+    //     else
+    //       if (subject == b)
+    //         true
+    //       else
+    //         subject == c) {
+    //     action
+    // }
+    //
+    // @return true if the conditions are equality checks of constants.
+    private fun matchConditions(condition: IrExpression, thenLabel: Label): Boolean {
+        if (condition is IrCall) {
+            // Leaf nodes: a lookup/table switch can be used if:
+            // 1. All branches are CALL 'EQEQ(Any?, Any?)': Boolean
+            // 2. All types of variables involved in comparison are in the same group of Char/Byte/Short/Int, String or enum.
+            // 3. All arg0 refer to the same value.
+            // 4. All arg1 are IrConst<*>.
+
+            if (condition.symbol != codegen.classCodegen.context.irBuiltIns.eqeqSymbol)
+                return false
+
+            // TODO: The check relies on the implementation that subject is always lowered to a local variable.
+            val candidate = condition.getValueArgument(0) as? IrGetValue ?: return false
+            subject = subject ?: candidate
+            if (candidate.symbol != subject!!.symbol)
+                return false
+            if (!candidate.type.isInt())
+                return false
+
+            val case = condition.getValueArgument(1) as? IrConst<*> ?: return false
+            if (case.kind != IrConstKind.Int)
+                return false
+
+            cases.add(Pair(case.value as Int, thenLabel))
+
+            return true
+        } else if (condition is IrWhen && condition.origin == IrStatementOrigin.WHEN_COMMA) {
+            // WHEN_COMMA should always be a Boolean.
+            assert(condition.type.isBoolean())
```

Comment:
Please move the comment to the assertion message and add context-specific info, e.g. something like
```
assert(condition.type.isBoolean() { "WHEN_COMMA should always be a Boolean: ${condition.dump()}" }
```

(We try to avoid asserts without messages in our project as much as is reasonable.)

Reply:
Done.